File: plugins/splash/plugin/ios/SplashOverlayView.swift
/// SplashOverlayView.swift
///
/// Full-screen UIView that mirrors the LaunchScreen.storyboard layout so the
/// runtime overlay can attach with no visual seam after the system splash
/// tears down. Auto-layout constraints keep the image and optional branding
/// pinned through rotation.
///
/// `fadeOut(durationMs:completion:)` runs on the main thread; callers must
/// hop to main before invoking.

import UIKit

final class DriftSplashOverlayView: UIView {

    private let imageView = UIImageView()
    private var brandingView: UIImageView?
    private let backgroundLayerView = UIView()

    init() {
        super.init(frame: .zero)
        translatesAutoresizingMaskIntoConstraints = false

        backgroundLayerView.translatesAutoresizingMaskIntoConstraints = false
        backgroundLayerView.backgroundColor = parseHex(DriftSplashConfig.backgroundColor)
        addSubview(backgroundLayerView)

        imageView.translatesAutoresizingMaskIntoConstraints = false
        imageView.contentMode = .scaleAspectFit
        imageView.image = UIImage(named: "DriftSplash")
        addSubview(imageView)

        NSLayoutConstraint.activate([
            backgroundLayerView.topAnchor.constraint(equalTo: topAnchor),
            backgroundLayerView.bottomAnchor.constraint(equalTo: bottomAnchor),
            backgroundLayerView.leadingAnchor.constraint(equalTo: leadingAnchor),
            backgroundLayerView.trailingAnchor.constraint(equalTo: trailingAnchor),

            imageView.centerXAnchor.constraint(equalTo: centerXAnchor),
            imageView.centerYAnchor.constraint(equalTo: centerYAnchor),
            imageView.widthAnchor.constraint(lessThanOrEqualTo: widthAnchor, multiplier: 0.6),
            imageView.heightAnchor.constraint(lessThanOrEqualTo: heightAnchor, multiplier: 0.6),
        ])

        installBrandingIfConfigured()

        // Observe dark-mode changes so the image and background swap live
        // while the splash is up.
        applyTraitColours()
    }

    required init?(coder: NSCoder) { fatalError("not supported") }

    private func installBrandingIfConfigured() {
        guard UIImage(named: "DriftSplashBranding") != nil else { return }
        let view = UIImageView(image: UIImage(named: "DriftSplashBranding"))
        view.translatesAutoresizingMaskIntoConstraints = false
        view.contentMode = .scaleAspectFit
        addSubview(view)
        brandingView = view

        let bottom = view.bottomAnchor.constraint(equalTo: safeAreaLayoutGuide.bottomAnchor, constant: -24)
        let height = view.heightAnchor.constraint(lessThanOrEqualToConstant: 80)
        NSLayoutConstraint.activate([bottom, height])

        switch DriftSplashConfig.brandingPosition {
        case "bottom_left":
            view.leadingAnchor.constraint(equalTo: safeAreaLayoutGuide.leadingAnchor, constant: 24).isActive = true
        case "bottom_right":
            view.trailingAnchor.constraint(equalTo: safeAreaLayoutGuide.trailingAnchor, constant: -24).isActive = true
        default:
            view.centerXAnchor.constraint(equalTo: centerXAnchor).isActive = true
        }
    }

    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        applyTraitColours()
    }

    private func applyTraitColours() {
        // Image assets named in the Assets.xcassets ship as a light/dark
        // pair under the same name when the user supplies a `dark:` variant;
        // UIImage(named:) handles the appearance match automatically.
        // Background colour is the only thing the runtime resolves itself.
        let bgHex = traitCollection.userInterfaceStyle == .dark && !DriftSplashConfig.darkBackgroundColor.isEmpty
            ? DriftSplashConfig.darkBackgroundColor
            : DriftSplashConfig.backgroundColor
        backgroundLayerView.backgroundColor = parseHex(bgHex)
    }

    func fadeOut(durationMs: Int, completion: @escaping () -> Void) {
        UIView.animate(
            withDuration: TimeInterval(durationMs) / 1000.0,
            delay: 0,
            options: [.curveEaseOut],
            animations: { self.alpha = 0 },
            completion: { _ in
                self.removeFromSuperview()
                completion()
            }
        )
    }

    private func parseHex(_ s: String) -> UIColor {
        // Accepts #RRGGBB and #RRGGBBAA. Validated at build time, so a
        // malformed input here means the build-time validator was bypassed.
        let trimmed = s.hasPrefix("#") ? String(s.dropFirst()) : s
        var value: UInt64 = 0
        guard Scanner(string: trimmed).scanHexInt64(&value) else { return .white }
        let r, g, b, a: CGFloat
        switch trimmed.count {
        case 6:
            r = CGFloat((value & 0xFF0000) >> 16) / 255
            g = CGFloat((value & 0x00FF00) >> 8) / 255
            b = CGFloat(value & 0x0000FF) / 255
            a = 1
        case 8:
            r = CGFloat((value & 0xFF000000) >> 24) / 255
            g = CGFloat((value & 0x00FF0000) >> 16) / 255
            b = CGFloat((value & 0x0000FF00) >> 8) / 255
            a = CGFloat(value & 0x000000FF) / 255
        default:
            return .white
        }
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }
}

File: plugins/splash/plugin/android/SplashOverlayView.kt
/**
 * SplashOverlayView.kt
 *
 * Full-screen FrameLayout that mirrors the LaunchTheme's launch_background
 * drawable so the runtime overlay can attach with no visual seam after the
 * system launch theme transitions to AppTheme. MATCH_PARENT layout params
 * keep the overlay glued to the activity's content area across rotation.
 *
 * `fadeOut(durationMs, onEnd)` runs on the main thread; callers must post
 * to a Handler before invoking.
 */
package com.drift.plugin.splash

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Context
import android.content.res.Configuration
import android.graphics.Color
import android.view.Gravity
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView

class DriftSplashOverlayView(context: Context) : FrameLayout(context) {

    private val imageView = ImageView(context).apply {
        scaleType = ImageView.ScaleType.FIT_CENTER
        val params = LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
        ).apply { gravity = Gravity.CENTER }
        layoutParams = params
    }

    private var brandingView: ImageView? = null

    init {
        layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT,
        )
        applyAppearance()
        imageView.setImageResource(resources.getIdentifier(
            "drift_splash", "drawable", context.packageName,
        ))
        addView(imageView)
        installBrandingIfPresent()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        applyAppearance()
    }

    private fun applyAppearance() {
        val dark = (resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) ==
            Configuration.UI_MODE_NIGHT_YES
        val bg = if (dark && DriftSplashConfig.DARK_BACKGROUND_COLOR.isNotEmpty())
            DriftSplashConfig.DARK_BACKGROUND_COLOR
        else
            DriftSplashConfig.BACKGROUND_COLOR
        setBackgroundColor(parseHex(bg))
    }

    private fun installBrandingIfPresent() {
        val brandingId = resources.getIdentifier(
            "drift_splash_branding", "drawable", context.packageName,
        )
        if (brandingId == 0) return
        val view = ImageView(context).apply {
            scaleType = ImageView.ScaleType.FIT_CENTER
            setImageResource(brandingId)
        }
        val gravity = when (DriftSplashConfig.BRANDING_POSITION) {
            "bottom_left" -> Gravity.BOTTOM or Gravity.START
            "bottom_right" -> Gravity.BOTTOM or Gravity.END
            else -> Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
        }
        view.layoutParams = LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT,
        ).apply {
            this.gravity = gravity
            bottomMargin = (resources.displayMetrics.density * 24).toInt()
            leftMargin = bottomMargin
            rightMargin = bottomMargin
        }
        addView(view)
        brandingView = view
    }

    fun fadeOut(durationMs: Int, onEnd: () -> Unit) {
        animate()
            .alpha(0f)
            .setDuration(durationMs.toLong())
            .setListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    (parent as? ViewGroup)?.removeView(this@DriftSplashOverlayView)
                    onEnd()
                }
            })
            .start()
    }

    private fun parseHex(s: String): Int {
        // Accepts #RRGGBB and #RRGGBBAA. Validated at build time.
        return try {
            Color.parseColor(s)
        } catch (_: IllegalArgumentException) {
            Color.WHITE
        }
    }
}
